src/modules/job-manager/plugins/dependency-singleton.c

```diff
@@ -0,0 +1,344 @@
+        || !(sctx->singletons = hola_create (flags)))
+        goto error;
+    return sctx;
+error:
+    singleton_ctx_destroy (sctx);
+    return NULL;
+}
+
+static int singleton_key_create (char *key,
+                                 size_t len,
+                                 uint32_t userid,
+                                 const char *name)
+{
+    if (snprintf (key, len, "%u:%s", userid, name) >= len)
+        return -1;
+    return 0;
+}
+
+static int singleton_list_push (flux_plugin_t *p,
+                                struct singleton_ctx *sctx,
+                                const char *key,
+                                flux_jobid_t id)
+{
+    struct singleton *s;
+
+    if (!(s = calloc (1, sizeof (*s)))
+        || flux_jobtap_job_aux_set (p, id, NULL, s, (flux_free_f) free) < 0) {
+        free (s);
+        return -1;
+    }
+    s->id = id;
+    if (!(hola_list_add_end (sctx->singletons, key, s)))
+        return -1;
+
+    return 0;
+}
+
+static struct singleton *singleton_list_pop (struct singleton_ctx *sctx,
+                                             const char *key)
+{
+    struct singleton *s = hola_list_first (sctx->singletons, key);
+    if (s) {
+        hola_list_delete (sctx->singletons,
+                          key,
+                          hola_list_cursor (sctx->singletons, key));
+    }
+    return s;
+}
+
+static size_t singleton_list_count (struct singleton_ctx *sctx,
+                                    const char *key)
+{
+    return hola_list_size (sctx->singletons, key);
+}
+
+/* Update singleton userid:name hash counts with value (1 or -1)
+ */
+static int singleton_count_update (struct singleton_ctx *sctx,
+                                   flux_plugin_t *p,
+                                   flux_plugin_arg_t *args,
+                                   int value)
+{
+    flux_jobid_t id;
+    struct singleton *s;
+    uint32_t userid;
+    const char *name = NULL;
+    char key [1024];
+    int count;
+
+    if (flux_plugin_arg_unpack (args,
+                                FLUX_PLUGIN_ARG_IN,
+                                "{s:I s:i s:{s:{s:{s?{s?s}}}}}",
+                                "id", &id,
+                                "userid", &userid,
+                                "jobspec",
+                                 "attributes",
+                                  "system",
+                                   "job",
+                                    "name", &name) < 0)
+        return -1;
+
+    /* Only track jobs with an explicit name */
+    if (name == NULL)
+        return 0;
+
+    if (singleton_key_create (key, sizeof (key), userid, name) < 0)
+        return -1;
+
+    if ((count = ptr2int (zhashx_lookup (sctx->counts, key)) + value) > 0
+        && count == singleton_list_count (sctx, key)) {
+        /*  All active jobs of this uid/name pair are waiting on a singleton
+         *  dependency. Pop the next job on the list and release it:
+         */
```

Comment:
I'm having a hard time reading the parens and side effects today.  It might be clearer to break this up like
```c
count = ptr2int (zhashx_lookup (sctx->counts, key));
count += value;
if (count > 0 && count == singleton_list_count (sctx, key)) {
```
Eh maybe I just need more :coffee:  You can take this one or leave it.

Reply:
Ok.. if going that far perhaps a function that abstracts fetching the current `count`. That will be self documenting.

Reply:
I just couldn't quickly spot if count was being incremented in the side effect assignment or if the test was that count + value > 0.

Reply:
New code looks like this, which I feel is much improved.
Better?

```c
    count = get_current_active_count (sctx, key) + value;
    if (count == singleton_list_count (sctx, key)
        && (s = singleton_list_pop (sctx, key))) {
        /*  All active jobs of this uid/name pair are waiting on a singleton
         *  dependency. Pop the next job on the list and release it:
         */
        flux_jobtap_dependency_remove (p, s->id, "singleton");
    }
    set_current_active_count (sctx, key, count);
```